src/workerd/jsg/jsvalue.h

```diff
@@ -0,0 +1,819 @@
+  operator v8::Local<v8::Value>() const { return inner; }
+  operator JsValue() const;
+  operator JsObject() const;
+  inline operator bool() const { return !inner.IsEmpty(); }
+
+  inline bool operator==(const JsValue& other) const KJ_WARN_UNUSED_RESULT {
+    return inner == other.inner;
+  }
+
+  inline uint32_t size() const KJ_WARN_UNUSED_RESULT { return inner->Length(); }
+
+  JsValue get(jsg::Lock& js, uint32_t i) const KJ_WARN_UNUSED_RESULT;
+
+  JsRef<JsArray> addRef(Lock& js) const KJ_WARN_UNUSED_RESULT;
+
+  inline explicit JsArray(v8::Local<v8::Value> inner) : JsArray(inner.As<v8::Array>()) {}
+  inline explicit JsArray(v8::Local<v8::Array> inner) : inner(inner) {
+    requireOnStack(this);
+  }
+
+  JSG_REQUIRE_STACK_ALLOCATION;
+
+private:
+  v8::Local<v8::Array> inner;
+  friend class JsValue;
+  friend class Lock;
+  template <typename TypeWrapper>
+  friend struct JsValueWrapper;
+  template <typename T> friend class JsRef;
+};
+
+#define V(Name) \
+  class Js##Name final { \
+  public: \
+    inline static bool is(const JsValue& value) KJ_WARN_UNUSED_RESULT { \
+      return value.inner->Is##Name(); \
+    } \
+    inline static bool is(v8::Local<v8::Value>& value) KJ_WARN_UNUSED_RESULT { \
+      return value->Is##Name(); \
+    } \
+    \
+    operator v8::Local<v8::Value>() const { return inner; } \
+    operator v8::Local<v8::Name>() const { return inner; } \
+    operator JsValue() const { return JsValue(inner.As<v8::Value>()); } \
+    inline operator bool() const { return !inner.IsEmpty(); } \
+    inline bool operator==(const JsValue& other) const KJ_WARN_UNUSED_RESULT { \
+      return inner == other.inner; \
+    } \
+    JsRef<Js##Name> addRef(Lock& js) const KJ_WARN_UNUSED_RESULT; \
+    inline explicit Js##Name(v8::Local<v8::Value> inner) \
+        : Js##Name(inner.template As<v8::Name>()) {} \
+    inline explicit Js##Name(v8::Local<v8::Name> inner) : inner(inner) { \
+      requireOnStack(this); \
+    } \
+    JSG_REQUIRE_STACK_ALLOCATION; \
+  private: \
+    v8::Local<v8::Name> inner; \
+    friend class JsValue; \
+    friend class JsObject; \
+    friend class Lock; \
+    template <typename TypeWrapper> \
+    friend struct JsValueWrapper; \
+    template <typename T> friend class JsRef; \
+  };
+
+  JS_TYPE_CLASSES(V)
+
+#undef V
+
+class JsObject final {
+public:
+  inline static bool is(const JsValue& value) KJ_WARN_UNUSED_RESULT {
+    return value.inner->IsObject();
+  }
+  inline static bool is(v8::Local<v8::Value>& value) KJ_WARN_UNUSED_RESULT {
+    return value->IsObject();
+  }
+
+  inline operator bool() const { return !inner.IsEmpty(); }
+  inline bool operator==(const JsValue& other) const {
+    return inner == other.inner;
+  }
+
+  operator JsValue() const;
+  operator JsMap() const;
+  operator v8::Local<v8::Value>() const { return inner; }
+  operator v8::Local<v8::Object>() const { return inner; }
+
+  template <typename T>
+  Promise<T> asPromise(Lock& js) const KJ_WARN_UNUSED_RESULT;
```

Comment:
Why is this downcast a named method instead of an `operator`? (I prefer a method but wondering if there's something special going on here.)

Can we have a general `tryDowncast<T>()` method instead of type-specfific methods? Note that you can define this template such that it's only enabled if `T` is actually a subclass according to V8, like:

```
template <JsSubclassOf<Object> T>
kj::Maybe<T> tryDowncast();
```

Where `JsSublcassOf` is a concept that checks if the corresponding V8 type is a subclass.

```
template <typename T, typename Base>
concept JsSubclassOf = kj::canConvert<typename T::V8Type*, typename Base::V8Type*>();
```

(This assumes each JsValue type has a member typedef liue `using V8Type = v8::Whatever;`)

(Disclaimer: I've never written a concept, might be getting it wrong.)

Reply:
I added this so we could have this exact conversation :-).  Nothing is using this currently and there are definitely a couple of ways we can handle these kinds of conversions. Important, these are not technically downcasts as `jsg::Promise<T>` is not assignable to `JsValue` at all. The reason this is a method is because of the preference to explicitly pass the `jsg::Lock&` argument to needed to create the `jsg::Promise<T>`... having it as a method is just better.

An argument could be made that we shouldn't have these kinds of transforms to other non-JsValue types at all, which I think also makes sense. We could have a variation of wrap on `TypeHandler` that accepts `JsValue` instead of a `v8::Local<T>` in the case someone wants to convert from a `JsValue` type to a more specific type.

Reply:
Oh, I missed the fact that this was returning `Promise` and not `JsPromise`. I guess this highlights the fact that there's going to be a lot of confusion here between these `Js-` types and the corresponding long-lived references.

I wonder if we want to move towards `jsg::Promise`, `jsg::Function`, etc. actually wrapping local handles instead of persistent handles? Usually, doing anything with a handle requires converting it to local anyway, so maybe this would actually be a better way...

Reply:
For now I have this where there is no `JsPromise` or `JsFunction`. We can use the existing `TypeHandler` API to convert these into appropriate `jsg::Promise<T>` and `jsg::Function<T>` instances.